# IBM Data Science Professional Certificate
# Capstone project
# Week 3

# Question 1. Scraping Wikipedia page for information on Toronto neighborhoods

Each major step of the process will be under a sub-heading. The answer to the questions will be marked down with a level 1 heading. Each step is in a different cell. Some cells are optional and will be marked as such.

## Installing libraries (optional)

This step installs the libraries necessary to run the notebook. If you already have these installed, you can and should skip running this cell. The libraries are:
* Beautifulsoup, this is used for parsing the html from wikipedia, although we will soon handover to pandas
* Geopy is used for geolocation
* Folium is used for map visualization

In [64]:
!conda install -c conda-forge beautifulsoup4 --yes

!conda install -c conda-forge geopy --yes

!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: done

# All requested packages already installed.

Solving environment: done

## Package Plan ##

  environment location: /Users/karimabbas/opt/anaconda3

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-2.0.0                |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          63 KB

The following packages will be UPDATED:

  geopy                                 1.22.0-pyh9f0ad1d_0 --> 2.0.0-pyh9f0ad1d_0



geopy-2.0.0          | 63 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
Solving environment: done

# All requested packages already installed.



## Importing

Always run this cell. We are importing all the libraries that will be used including beautifulsoup, numpy, pandas, folium, and sklearn among others. If any of the import conditions yield an error, you need to install the corresponding library first using !pip. Within the rest of the notebook, we will import additional libraries and components of libraries as needed. Make sure to run these cells as well.

In [65]:
import numpy as np
import pandas as pd

import requests

import requests
from pandas.io.json import json_normalize
import json

import matplotlib.cm as cm

import matplotlib.colors as colors


from bs4 import BeautifulSoup

from geopy.geocoders import Nominatim

import folium


import sklearn
from sklearn.cluster import KMeans

## Create a soup object

In this step, we create a handle for the text contents of the link to the html page and pass it to beautiful soup. This creates an object of type bs4.BeautifulSoup. If you try printing its contents with the .text() method, you will find a very raw form of the html on the page.

In [66]:
url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(url, 'html.parser')

This code segment creates empty lists that will later be populated by data from the table

In [67]:
postalcodes = []
boroughs = []
neighborhoods = []


The cell below performs the parsing. Comments indicate what is happening.

In [68]:
# This loop will scour the soup object line by line to find where the first table starts
# It is actually looking for an html tag rather than the specific string
# It then looks for and returns all elements in the table with tag <tr> which indicates a table row
# We then extract any entry with a tage <td> which indicates a normal table cell entry
for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    #print(row)
    #print(cells)
    if(len(cells) > 0):
        # The above check avoids scouring formatting and empty rows
        #print(cells[0].text)
        postalcodes.append(cells[0].text.replace("\n",""))
        boroughs.append(cells[1].text.replace("\n",""))
        neighborhoods.append(cells[2].text.replace("\n",""))
        # We populate the three empty lists with the elements of the table
        # We have to use the replace method because for some reason each entry includes an extra carriage return
        # Without the replace method, my table has an extra \n literal added to the string
        
#print(postalCodeList)
#print(boroughList)
#print(neighborhoodList)

## Forming the main data frame with Toronto geographical information

The upcoming cells will create the main geographical information data frame we will be using for the rest of the problem. In the first step we create a dictionary of labels and corresponding values for postalcodes, boroughs, and neighborhoods. There are three dictionary keys and the values corresponding to the three keys are the lists we populated above using the soup method.

While a dictionary may not seem like an immediately suitable data structure here, there is a reason we use it. If you try printing the dictionary (there is a commented line to do this), you will find that the entries look counter intuitive and not very instructive. There is a very short string key corresponding to an extremely long value list. However, the DataFrame.from_dict() method is why we are doing this. The method takes in the dictionary and creates a data frame where the keys are the column headings and the lists are the values.

We print a header for the dataframe as a quick check that we did not mess up the formatting. This is something we will be doing after every significant step that changes our main dataframe.

In [69]:
# We create a dictionary of three lists we created. This allows us to use a special pandas method to create the frame

n_dict = {'PostalCode': postalcodes,
                      'Borough': boroughs,
                      'Neighborhood': neighborhoods}
#print(n_dict)
df = pd.DataFrame.from_dict(n_dict)
# As a sanity check print a long header
df.head(10)

PostalCode           Borough                                 Neighborhood
0        M1A      Not assigned                                 Not assigned
1        M2A      Not assigned                                 Not assigned
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7        M8A      Not assigned                                 Not assigned
8        M9A         Etobicoke      Islington Avenue, Humber Valley Village
9        M1B       Scarborough                               Malvern, Rouge

Next is inital cleanup. We are required to remove all entries which do not have a Borough value. This is indicated by the value 'Not assigned' in the frame. The default pandas method will act on field values of 'na'. However, we can create a new frame where the 'Not assigned' fields are dropped. This cannot be done in place, so we create a new frame called toronto_df_nona. As always, we print a header to check, although in this case this is not an exhaustive check.

In [70]:
toronto_df_nona = df[df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_nona.head(10)

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5        M9A         Etobicoke      Islington Avenue, Humber Valley Village
6        M1B       Scarborough                               Malvern, Rouge
7        M3B        North York                                    Don Mills
8        M4B         East York              Parkview Hill, Woodbine Gardens
9        M5B  Downtown Toronto                     Garden District, Ryerson

In [71]:
toronto_df_grouped = toronto_df_nona.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [72]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'
toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']
toronto_df_grouped[na_neigh_rows]
toronto_df_grouped.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

In [73]:
toronto_df_cleaned = toronto_df_grouped
toronto_df_cleaned.shape

(103, 3)

As seen above the frame has 103 entries. There are three columns in the frame for postal code, borough, and neighborhood. The key of the frame is PostalCode, although we have not yet specified this.

___
# Question 2. Getting coordinates and merging the frames

Either download the file from the attached URL then load it using .read_csv() or uncommet the !wget commands and obtain it from the provided link. As always, the first thing we do is get to know the data. We do this by getting familiar with its shape and taking a look at the first few records. You can either use the read_csv() method with a local file or uncomment !wget to obtain online.

In [74]:
#!pip install wget
import wget
#!wget -q -O "toronto_coordinates.csv" http://cocl.us/Geospatial_data
#!wget --quiet https://cocl.us/sanfran_geojson -O sf_geo.json    
coordinates = pd.read_csv('Geospatial_Coordinates.csv')

In [75]:
print(coordinates.shape)
coordinates.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [76]:

toronto_df_temp = toronto_df_cleaned.set_index('PostalCode')
coors_temp = coordinates.set_index('Postal Code')
toronto_df_coors = pd.concat([toronto_df_temp, coors_temp], axis=1, join='inner')

In [77]:
toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

In [78]:
print(toronto_df_coors.shape)
toronto_df_coors.head()

(103, 5)


PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Above, we define indices for both frames as the PostalCode column. This allows us to merge the frame containing name information with the frame containing latitude and longitude by the common field: postal code. We use the concat method although the merge method would also work. As always, we use shape and head to get a first look at the data. This always works as a sanity check.

## First reduction

This section reduces the number of entries in the frame by only considering neighborhoods with the name Toronto in them. We end up with 39 entries corresponding to 39 distinct postal codes, all beginning with "M".

In [79]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
#toronto_central_df.head()

(39, 5)


In [80]:
toronto_central_df.head(39)

PostalCode           Borough  \
0         M4E      East Toronto   
1         M4K      East Toronto   
2         M4L      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
5         M4P   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
10        M4W  Downtown Toronto   
11        M4X  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
14        M5B  Downtown Toronto   
15        M5C  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
18        M5H  Downtown Toronto   
19        M5J  Downtown Toronto   
20        M5K  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
24        M5R   Central Toronto   
25        M5S  Downtown Toronto   
26        M5T  Downtown Toronto   
27        M5V  Downtown Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
30        M6G  Downtown Toronto   
31        M6H      West Toronto   
32        M6J      West Toronto   
33        M6K      West Toronto   
34        M6P      West Toronto   
35        M6R      West Toronto   
36        M6S      West Toronto   
37        M7A  Downtown Toronto   
38        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                        The Danforth West, Riverdale  43.679557 -79.352188  
2                      India Bazaar, The Beaches West  43.668999 -79.315572  
3                                     Studio District  43.659526 -79.340923  
4                                       Lawrence Park  43.728020 -79.388790  
5                                    Davisville North  43.712751 -79.390197  
6                  North Toronto West,  Lawrence Park  43.715383 -79.405678  
7                                          Davisville  43.704324 -79.388790  
8                         Moore Park, Summerhill East  43.689574 -79.383160  
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
10                                           Rosedale  43.679563 -79.377529  
11                        St. James Town, Cabbagetown  43.667967 -79.367675  
12                               Church and Wellesley  43.665860 -79.383160  
13                          Regent Park, Harbourfront  43.654260 -79.360636  
14                           Garden District, Ryerson  43.657162 -79.378937  
15                                     St. James Town  43.651494 -79.375418  
16                                        Berczy Park  43.644771 -79.373306  
17                                 Central Bay Street  43.657952 -79.387383  
18                           Richmond, Adelaide, King  43.650571 -79.384568  
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
22                                           Roselawn  43.711695 -79.416936  
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                     University of Toronto, Harbord  43.662696 -79.400049  
26          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
28                                     Stn A PO Boxes  43.646435 -79.374846  
29             First Canadian Place, Underground city  43.648429 -79.382280  
30                                           Christie  43.669542 -79.422564  
31                       Dufferin, Dovercourt Village  43.669005 -79.442259  
32                          

# Question 3. Clustering neighborhoods of Toronto

The code below emulates what we did with New York in the lab. First, we draw a map of Toronto and mark neighborhoods. Then we define a function to pull venue information from Foursquare. We then call this function once for one location as a test and then for all neighborhoods in application. We merge the information with the data frame, then perform K-means clustering with four clusters. The map at the very end shows the results of clustering.

In [81]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


In [82]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [83]:
CLIENT_ID = '20BDBJM2RIGFPVT5L4PL2UW3TXRHVCKGU1GHCPMONFR4IHZH' # your Foursquare ID
CLIENT_SECRET = 'HKVHQUZEU02L52EMRWBFGVQEBAGHSIC0YCO1FTWS2BBNI2HY' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 20BDBJM2RIGFPVT5L4PL2UW3TXRHVCKGU1GHCPMONFR4IHZH
CLIENT_SECRET:HKVHQUZEU02L52EMRWBFGVQEBAGHSIC0YCO1FTWS2BBNI2HY


In [84]:
#manhattan_data.loc[0, 'Neighborhood']

In [85]:
neighborhood_latitude = toronto_central_df.loc[16, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_central_df.loc[16, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_central_df.loc[16, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Berczy Park are 43.644770799999996, -79.3733064.


In [86]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=20BDBJM2RIGFPVT5L4PL2UW3TXRHVCKGU1GHCPMONFR4IHZH&client_secret=HKVHQUZEU02L52EMRWBFGVQEBAGHSIC0YCO1FTWS2BBNI2HY&v=20180605&ll=43.644770799999996,-79.3733064&radius=500&limit=100'

In [87]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ef8813e042e2d70f180f110'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Downtown Toronto',
  'headerFullLocation': 'Downtown Toronto, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 59,
  'suggestedBounds': {'ne': {'lat': 43.6492708045, 'lng': -79.36709938085544},
   'sw': {'lat': 43.640270795499994, 'lng': -79.37951341914457}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4aeb719af964a52020c221e3',
       'name': 'LCBO',
       'location': {'address': '2 Cooper St',
        'crossStreet': 'at Queens Quay E',
        'lat': 43.64294379917171,
        'lng': -79.37243989044406,
        'labeledLatLngs': [{'label'

In [88]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [89]:
#venues = results['response']['groups'][0]['items']
    
#nearby_venues = json_normalize(venues) # flatten JSON

## filter columns
#filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
#nearby_venues =nearby_venues.loc[:, filtered_columns]

## filter the category for each row
#nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

## clean columns
#nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

#nearby_venues.head()

In [90]:
#print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

In [91]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        print(results)
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [92]:
LIMIT=100
radius=500
toronto_venues = getNearbyVenues(names=toronto_central_df['Neighborhood'],
                                   latitudes=toronto_central_df['Latitude'],
                                   longitudes=toronto_central_df['Longitude']
                                  )

The Beaches
[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4bd461bc77b29c74a07d9282', 'name': 'Glen Manor Ravine', 'location': {'address': 'Glen Manor', 'crossStreet': 'Queen St.', 'lat': 43.67682094413784, 'lng': -79.29394208780985, 'labeledLatLngs': [{'label': 'display', 'lat': 43.67682094413784, 'lng': -79.29394208780985}], 'distance': 89, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['Glen Manor (Queen St.)', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d159941735', 'name': 'Trail', 'pluralName': 'Trails', 'shortName': 'Trail', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/hikingtrail_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4bd461bc77b29c74a07d9282-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'ge

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ade390ff964a5200e7421e3', 'name': 'System Fitness', 'location': {'address': '1661 Queen St East', 'crossStreet': 'at Kingston Rd', 'lat': 43.667171452103204, 'lng': -79.31273345707406, 'labeledLatLngs': [{'label': 'display', 'lat': 43.667171452103204, 'lng': -79.31273345707406}], 'distance': 305, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['1661 Queen St East (at Kingston Rd)', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d176941735', 'name': 'Gym', 'pluralName': 'Gyms', 'shortName': 'Gym', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/gym_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4ade390ff964a5200e7421e3-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general'

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ad7e958f964a520001021e3', 'name': "Ed's Real Scoop", 'location': {'address': '920 Queen St. E', 'crossStreet': 'btwn Logan Ave. & Morse St.', 'lat': 43.660655832455014, 'lng': -79.3420187548006, 'labeledLatLngs': [{'label': 'display', 'lat': 43.660655832455014, 'lng': -79.3420187548006}], 'distance': 153, 'postalCode': 'M4M 1J5', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['920 Queen St. E (btwn Logan Ave. & Morse St.)', 'Toronto ON M4M 1J5', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1c9941735', 'name': 'Ice Cream Shop', 'pluralName': 'Ice Cream Shops', 'shortName': 'Ice Cream', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/icecream_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4ad7e958f964a520001021e3-0'}, {'reaso

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '50e6da19e4b0d8a78a0e9794', 'name': 'Lawrence Park Ravine', 'location': {'address': '3055 Yonge Street', 'crossStreet': 'Lawrence Avenue East', 'lat': 43.72696303913755, 'lng': -79.39438246708775, 'labeledLatLngs': [{'label': 'display', 'lat': 43.72696303913755, 'lng': -79.39438246708775}], 'distance': 465, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['3055 Yonge Street (Lawrence Avenue East)', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d163941735', 'name': 'Park', 'pluralName': 'Parks', 'shortName': 'Park', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-50e6da19e4b0d8a78a0e9794-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is pop

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '51606062e4b0878cf540f4a2', 'name': 'Barreworks', 'location': {'address': '2576 Yonge St', 'lat': 43.71407030751952, 'lng': -79.40010911522093, 'labeledLatLngs': [{'label': 'display', 'lat': 43.71407030751952, 'lng': -79.40010911522093}], 'distance': 471, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['2576 Yonge St', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d102941735', 'name': 'Yoga Studio', 'pluralName': 'Yoga Studios', 'shortName': 'Yoga Studio', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/gym_yogastudio_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-51606062e4b0878cf540f4a2-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteraction

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4b939ca4f964a520344a34e3', 'name': 'Thobors Boulangerie Patisserie Café', 'location': {'address': '627 Mt Pleasant Rd.', 'crossStreet': 'btwn Manor Rd. and Hillsdale Ave.', 'lat': 43.704513877453266, 'lng': -79.38861602551758, 'labeledLatLngs': [{'label': 'display', 'lat': 43.704513877453266, 'lng': -79.38861602551758}], 'distance': 25, 'postalCode': 'M4S 2M9', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['627 Mt Pleasant Rd. (btwn Manor Rd. and Hillsdale Ave.)', 'Toronto ON M4S 2M9', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d16d941735', 'name': 'Café', 'pluralName': 'Cafés', 'shortName': 'Café', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/cafe_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4b939ca4f964a520344a34e3-

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4db7452da86ed8d46c7117f9', 'name': 'Loring-Wyle Parkette', 'location': {'address': '276 St. Clair Avenue East', 'lat': 43.690270427217385, 'lng': -79.3834375880377, 'labeledLatLngs': [{'label': 'display', 'lat': 43.690270427217385, 'lng': -79.3834375880377}], 'distance': 80, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['276 St. Clair Avenue East', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d163941735', 'name': 'Park', 'pluralName': 'Parks', 'shortName': 'Park', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4db7452da86ed8d46c7117f9-0'}]
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
[{'reasons': {'count': 0, 'items': [{'summary

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4aff2d47f964a520743522e3', 'name': 'Rosedale Park', 'location': {'address': '38 Scholfield Ave.', 'crossStreet': 'at Edgar Ave.', 'lat': 43.68232820227814, 'lng': -79.37893434347683, 'labeledLatLngs': [{'label': 'display', 'lat': 43.68232820227814, 'lng': -79.37893434347683}], 'distance': 327, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['38 Scholfield Ave. (at Edgar Ave.)', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1e7941735', 'name': 'Playground', 'pluralName': 'Playgrounds', 'shortName': 'Playground', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/playground_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4aff2d47f964a520743522e3-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot i

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '51a11515498ef97667980b8c', 'name': 'DanceLifeX Centre', 'location': {'address': '619 Yonge Street', 'crossStreet': 'btwn Isabella St & Gloucester St', 'lat': 43.66695583448771, 'lng': -79.3852972984314, 'labeledLatLngs': [{'label': 'display', 'lat': 43.66695583448771, 'lng': -79.3852972984314}], 'distance': 210, 'postalCode': 'M4Y 1Z5', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['619 Yonge Street (btwn Isabella St & Gloucester St)', 'Toronto ON M4Y 1Z5', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d134941735', 'name': 'Dance Studio', 'pluralName': 'Dance Studios', 'shortName': 'Dance Studio', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/arts_entertainment/performingarts_dancestudio_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}, 'venuePage': {

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '54ea41ad498e9a11e9e13308', 'name': 'Roselle Desserts', 'location': {'address': '362 King St E', 'crossStreet': 'Trinity St', 'lat': 43.653446723052674, 'lng': -79.3620167174383, 'labeledLatLngs': [{'label': 'display', 'lat': 43.653446723052674, 'lng': -79.3620167174383}], 'distance': 143, 'postalCode': 'M5A 1K9', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['362 King St E (Trinity St)', 'Toronto ON M5A 1K9', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d16a941735', 'name': 'Bakery', 'pluralName': 'Bakeries', 'shortName': 'Bakery', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/bakery_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-54ea41ad498e9a11e9e13308-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is pop

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ad4c062f964a5200bf820e3', 'name': 'Silver Snail Comics', 'location': {'address': '329 Yonge St', 'crossStreet': 'at Dundas St E', 'lat': 43.65703137958407, 'lng': -79.38140310220501, 'labeledLatLngs': [{'label': 'display', 'lat': 43.65703137958407, 'lng': -79.38140310220501}], 'distance': 199, 'postalCode': 'M5B 1R7', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['329 Yonge St (at Dundas St E)', 'Toronto ON M5B 1R7', 'Canada']}, 'categories': [{'id': '52f2ab2ebcbc57f1066b8b18', 'name': 'Comic Shop', 'pluralName': 'Comic Shops', 'shortName': 'Comic Shop', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/comic_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4ad4c062f964a5200bf820e3-0'}, {'reasons': {'count': 0, 'items': [{'summary

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '574ad72238fa943556d93b8e', 'name': 'Gyu-Kaku Japanese BBQ', 'location': {'address': '81 Church St', 'crossStreet': 'at Adelaide St E', 'lat': 43.651422275497914, 'lng': -79.37504693687086, 'labeledLatLngs': [{'label': 'display', 'lat': 43.651422275497914, 'lng': -79.37504693687086}], 'distance': 30, 'postalCode': 'M5C 2G2', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['81 Church St (at Adelaide St E)', 'Toronto ON M5C 2G2', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d111941735', 'name': 'Japanese Restaurant', 'pluralName': 'Japanese Restaurants', 'shortName': 'Japanese', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/japanese_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-574ad72238fa943556d93b8e-0'}, {'reasons': {'count

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4aeb719af964a52020c221e3', 'name': 'LCBO', 'location': {'address': '2 Cooper St', 'crossStreet': 'at Queens Quay E', 'lat': 43.64294379917171, 'lng': -79.37243989044406, 'labeledLatLngs': [{'label': 'display', 'lat': 43.64294379917171, 'lng': -79.37243989044406}], 'distance': 215, 'postalCode': 'M5E 1A4', 'cc': 'CA', 'neighborhood': 'Downtown Toronto', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['2 Cooper St (at Queens Quay E)', 'Toronto ON M5E 1A4', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d186941735', 'name': 'Liquor Store', 'pluralName': 'Liquor Stores', 'shortName': 'Liquor Store', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/food_liquor_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4aeb719af964a52020c221e3-0'}, {'reasons

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '537d4d6d498ec171ba22e7fe', 'name': "Jimmy's Coffee", 'location': {'address': '82 Gerrard Street W', 'crossStreet': 'Gerrard & LaPlante', 'lat': 43.65842123574496, 'lng': -79.38561319551111, 'labeledLatLngs': [{'label': 'display', 'lat': 43.65842123574496, 'lng': -79.38561319551111}], 'distance': 151, 'postalCode': 'M5G 1Z4', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['82 Gerrard Street W (Gerrard & LaPlante)', 'Toronto ON M5G 1Z4', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1e0931735', 'name': 'Coffee Shop', 'pluralName': 'Coffee Shops', 'shortName': 'Coffee Shop', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-537d4d6d498ec171ba22e7fe-0'}, {'reasons': {'count'

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ad4c05ef964a520a6f620e3', 'name': 'Nathan Phillips Square', 'location': {'address': '100 Queen St W', 'crossStreet': 'at Bay St', 'lat': 43.65227047322295, 'lng': -79.38351631164551, 'labeledLatLngs': [{'label': 'display', 'lat': 43.65227047322295, 'lng': -79.38351631164551}], 'distance': 207, 'postalCode': 'M5H 2N1', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['100 Queen St W (at Bay St)', 'Toronto ON M5H 2N1', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d164941735', 'name': 'Plaza', 'pluralName': 'Plazas', 'shortName': 'Plaza', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/plaza_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4ad4c05ef964a520a6f620e3-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This 

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4bfaa3494a67c928d08528cf', 'name': 'Harbourfront', 'location': {'lat': 43.639525632239106, 'lng': -79.38068838052389, 'labeledLatLngs': [{'label': 'display', 'lat': 43.639525632239106, 'lng': -79.38068838052389}], 'distance': 167, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['Toronto ON', 'Canada']}, 'categories': [{'id': '4f2a25ac4b909258e854f55f', 'name': 'Neighborhood', 'pluralName': 'Neighborhoods', 'shortName': 'Neighborhood', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/neighborhood_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4bfaa3494a67c928d08528cf-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4b

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ad4c05df964a52059f620e3', 'name': 'Canoe', 'location': {'address': '66 Wellington St West', 'crossStreet': 'at Bay Street', 'lat': 43.647452066183476, 'lng': -79.38132001815676, 'labeledLatLngs': [{'label': 'display', 'lat': 43.647452066183476, 'lng': -79.38132001815676}], 'distance': 36, 'postalCode': 'M5K 1H6', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['66 Wellington St West (at Bay Street)', 'Toronto ON M5K 1H6', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1c4941735', 'name': 'Restaurant', 'pluralName': 'Restaurants', 'shortName': 'Restaurant', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/default_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}, 'venuePage': {'id': '40552945'}}, 'referralId': 'e-0-4ad4c05df964a52059f620e3-0'}, {'reaso

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ea43642be7ba4918efc5afa', 'name': 'Adelaide Club Toronto', 'location': {'address': '1 First Canadian Place', 'lat': 43.64927944291712, 'lng': -79.38192123284551, 'labeledLatLngs': [{'label': 'display', 'lat': 43.64927944291712, 'lng': -79.38192123284551}], 'distance': 207, 'postalCode': 'M5X 1C8', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['1 First Canadian Place', 'Toronto ON M5X 1C8', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d175941735', 'name': 'Gym / Fitness Center', 'pluralName': 'Gyms or Fitness Centers', 'shortName': 'Gym / Fitness', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/gym_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4ea43642be7ba4918efc5afa-0'}, {'reasons': {'count': 0, 'items': [{'summary':

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '5ae8cfe98496ca002c65d83a', 'name': 'Ceiling Champions', 'location': {'address': '505 The West Mall Suite #207', 'lat': 43.71389073142895, 'lng': -79.42070245742796, 'labeledLatLngs': [{'label': 'display', 'lat': 43.71389073142895, 'lng': -79.42070245742796}], 'distance': 389, 'postalCode': 'M9C 1G2', 'cc': 'CA', 'city': 'Etobicoke', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['505 The West Mall Suite #207', 'Etobicoke ON M9C 1G2', 'Canada']}, 'categories': [{'id': '545419b1498ea6ccd0202f58', 'name': 'Home Service', 'pluralName': 'Home Services & Repairs', 'shortName': 'Home Services', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/hardware_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-5ae8cfe98496ca002c65d83a-0'}, {'reasons': {'count': 0, 'items': [{'sum

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '5362c366498e602fbe1db395', 'name': 'Yasu', 'location': {'address': '81 Harbord St.', 'lat': 43.66283719650635, 'lng': -79.40321739973975, 'labeledLatLngs': [{'label': 'display', 'lat': 43.66283719650635, 'lng': -79.40321739973975}], 'distance': 255, 'postalCode': 'M5S 1G4', 'cc': 'CA', 'neighborhood': 'Harbord Village, Toronto, ON', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['81 Harbord St.', 'Toronto ON M5S 1G4', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d111941735', 'name': 'Japanese Restaurant', 'pluralName': 'Japanese Restaurants', 'shortName': 'Japanese', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/japanese_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-5362c366498e602fbe1db395-0'}, {'reasons': {'count': 0, 'items': [{'su

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '50427a03e4b08d9f5931f593', 'name': 'Seven Lives - Tacos y Mariscos', 'location': {'address': '69 Kensington Ave', 'crossStreet': 'Baldwin St', 'lat': 43.65441817160925, 'lng': -79.40054467846261, 'labeledLatLngs': [{'label': 'display', 'lat': 43.65441817160925, 'lng': -79.40054467846261}], 'distance': 140, 'postalCode': 'M5T 2L6', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['69 Kensington Ave (Baldwin St)', 'Toronto ON M5T 2L6', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1c1941735', 'name': 'Mexican Restaurant', 'pluralName': 'Mexican Restaurants', 'shortName': 'Mexican', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/mexican_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}, 'venuePage': {'id': '36738371'}}, 'referralId': 'e-0-50427a03e4b08d

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ad8df54f964a520881521e3', 'name': 'Billy Bishop Toronto City Airport (YTZ) (Billy Bishop Toronto City Airport)', 'location': {'address': 'Toronto Island', 'lat': 43.63168259661481, 'lng': -79.3960334124689, 'distance': 331, 'postalCode': 'M5J 1B7', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['Toronto Island', 'Toronto ON M5J 1B7', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1ed931735', 'name': 'Airport', 'pluralName': 'Airports', 'shortName': 'Airport', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/travel/airport_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4ad8df54f964a520881521e3-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'v

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '5b5bca904aa3f8002c97f85d', 'name': 'Fresh On Front', 'location': {'address': '47 Front Street E', 'crossStreet': 'Church Street', 'lat': 43.64781466405137, 'lng': -79.3744527618627, 'labeledLatLngs': [{'label': 'display', 'lat': 43.64781466405137, 'lng': -79.3744527618627}], 'distance': 156, 'postalCode': 'M5E 1B3', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['47 Front Street E (Church Street)', 'Toronto ON M5E 1B3', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1d3941735', 'name': 'Vegetarian / Vegan Restaurant', 'pluralName': 'Vegetarian / Vegan Restaurants', 'shortName': 'Vegetarian / Vegan', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/vegetarian_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-5b5bca904aa3f8002c97f85

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4ea43642be7ba4918efc5afa', 'name': 'Adelaide Club Toronto', 'location': {'address': '1 First Canadian Place', 'lat': 43.64927944291712, 'lng': -79.38192123284551, 'labeledLatLngs': [{'label': 'display', 'lat': 43.64927944291712, 'lng': -79.38192123284551}], 'distance': 98, 'postalCode': 'M5X 1C8', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['1 First Canadian Place', 'Toronto ON M5X 1C8', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d175941735', 'name': 'Gym / Fitness Center', 'pluralName': 'Gyms or Fitness Centers', 'shortName': 'Gym / Fitness', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/gym_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4ea43642be7ba4918efc5afa-0'}, {'reasons': {'count': 0, 'items': [{'summary': 

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4adcfd7cf964a5203e6321e3', 'name': 'Fiesta Farms', 'location': {'address': '200 Christie St', 'crossStreet': 'at Essex St', 'lat': 43.66847077052224, 'lng': -79.42048512748114, 'labeledLatLngs': [{'label': 'display', 'lat': 43.66847077052224, 'lng': -79.42048512748114}], 'distance': 205, 'postalCode': 'M6G 3B6', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['200 Christie St (at Essex St)', 'Toronto ON M6G 3B6', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d118951735', 'name': 'Grocery Store', 'pluralName': 'Grocery Stores', 'shortName': 'Grocery Store', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/shops/food_grocery_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}, 'venuePage': {'id': '56848730'}}, 'referralId': 'e-0-4adcfd7cf964a5203e6321e3-0'}, {'

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '5753753b498eeb535c53aed5', 'name': 'The Greater Good Bar', 'location': {'address': '229 Geary St', 'crossStreet': 'at Dufferin St', 'lat': 43.669409, 'lng': -79.439267, 'labeledLatLngs': [{'label': 'display', 'lat': 43.669409, 'lng': -79.439267}], 'distance': 245, 'postalCode': 'M6H 2C1', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['229 Geary St (at Dufferin St)', 'Toronto ON M6H 2C1', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d116941735', 'name': 'Bar', 'pluralName': 'Bars', 'shortName': 'Bar', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/nightlife/pub_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-5753753b498eeb535c53aed5-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'rea

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4f7891c7e4b0b9643b73e08d', 'name': 'Bellwoods Brewery', 'location': {'address': '124 Ossington Ave', 'crossStreet': 'at Argyle St', 'lat': 43.647097254598236, 'lng': -79.41995537873463, 'labeledLatLngs': [{'label': 'display', 'lat': 43.647097254598236, 'lng': -79.41995537873463}], 'distance': 93, 'postalCode': 'M6J 2Z5', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['124 Ossington Ave (at Argyle St)', 'Toronto ON M6J 2Z5', 'Canada']}, 'categories': [{'id': '50327c8591d4c4b30a586d5d', 'name': 'Brewery', 'pluralName': 'Breweries', 'shortName': 'Brewery', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/brewery_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4f7891c7e4b0b9643b73e08d-0'}, {'reasons': {'count': 0, 'items': [{'summary':

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4f54ef6ce4b0929810978bb6', 'name': 'Reebok Crossfit Liberty Village', 'location': {'address': 'Liberty Village', 'lat': 43.637035591932005, 'lng': -79.42480199455974, 'labeledLatLngs': [{'label': 'display', 'lat': 43.637035591932005, 'lng': -79.42480199455974}], 'distance': 273, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['Liberty Village', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d176941735', 'name': 'Gym', 'pluralName': 'Gyms', 'shortName': 'Gym', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/gym_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4f54ef6ce4b0929810978bb6-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReaso

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4bd30e129854d13acdbefc4d', 'name': 'Lithuania Park', 'location': {'address': '155 Oakmount Rd', 'crossStreet': 'at Glenlake & Keele', 'lat': 43.658666653470426, 'lng': -79.46303844451904, 'labeledLatLngs': [{'label': 'display', 'lat': 43.658666653470426, 'lng': -79.46303844451904}], 'distance': 355, 'postalCode': 'M1P 4N7', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['155 Oakmount Rd (at Glenlake & Keele)', 'Toronto ON M1P 4N7', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d163941735', 'name': 'Park', 'pluralName': 'Parks', 'shortName': 'Park', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4bd30e129854d13acdbefc4d-0'}, {'reasons': {'count': 0, 'items': [{'summ

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4b757192f964a520d30c2ee3', 'name': 'Offleash Dog Trail - High Park', 'location': {'address': 'Centre Rd', 'crossStreet': 'btw High Park Blvd & Bloor', 'lat': 43.64548543069719, 'lng': -79.45874691009521, 'labeledLatLngs': [{'label': 'display', 'lat': 43.64548543069719, 'lng': -79.45874691009521}], 'distance': 433, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['Centre Rd (btw High Park Blvd & Bloor)', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1e5941735', 'name': 'Dog Run', 'pluralName': 'Dog Runs', 'shortName': 'Dog Run', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/dogrun_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4b757192f964a520d30c2ee3-0'}, {'reasons': {'count': 0, 'items': [{'summary': 

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4b15aef7f964a52074b223e3', 'name': 'Asa Sushi', 'location': {'address': '18 Jane St.', 'crossStreet': 'Bloor', 'lat': 43.64990164450559, 'lng': -79.4846109228881, 'labeledLatLngs': [{'label': 'display', 'lat': 43.64990164450559, 'lng': -79.4846109228881}], 'distance': 186, 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['18 Jane St. (Bloor)', 'Toronto ON', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d1d2941735', 'name': 'Sushi Restaurant', 'pluralName': 'Sushi Restaurants', 'shortName': 'Sushi', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/sushi_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4b15aef7f964a52074b223e3-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName':

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '4b9d206bf964a520e69136e3', 'name': "Queen's Park", 'location': {'address': 'University Ave.', 'crossStreet': 'at Wellesley Ave.', 'lat': 43.66394609897775, 'lng': -79.39217952520835, 'labeledLatLngs': [{'label': 'display', 'lat': 43.66394609897775, 'lng': -79.39217952520835}], 'distance': 283, 'postalCode': 'M5R 2E8', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['University Ave. (at Wellesley Ave.)', 'Toronto ON M5R 2E8', 'Canada']}, 'categories': [{'id': '4bf58dd8d48988d163941735', 'name': 'Park', 'pluralName': 'Parks', 'shortName': 'Park', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-4b9d206bf964a520e69136e3-0'}, {'reasons': {'count': 0, 'items': [{'summary': 'T

[{'reasons': {'count': 0, 'items': [{'summary': 'This spot is popular', 'type': 'general', 'reasonName': 'globalInteractionReason'}]}, 'venue': {'id': '58d4650e2f91cb5ccec03447', 'name': 'Rorschach Brewing Co.', 'location': {'address': '1001 Eastern Ave', 'crossStreet': 'Woodfield Rd', 'lat': 43.6634831695922, 'lng': -79.31982368639481, 'labeledLatLngs': [{'label': 'display', 'lat': 43.6634831695922, 'lng': -79.31982368639481}], 'distance': 162, 'postalCode': 'M4L 1A8', 'cc': 'CA', 'city': 'Toronto', 'state': 'ON', 'country': 'Canada', 'formattedAddress': ['1001 Eastern Ave (Woodfield Rd)', 'Toronto ON M4L 1A8', 'Canada']}, 'categories': [{'id': '50327c8591d4c4b30a586d5d', 'name': 'Brewery', 'pluralName': 'Breweries', 'shortName': 'Brewery', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/brewery_', 'suffix': '.png'}, 'primary': True}], 'photos': {'count': 0, 'groups': []}}, 'referralId': 'e-0-58d4650e2f91cb5ccec03447-0'}, {'reasons': {'count': 0, 'items': [{'summary': 

In [93]:
print(toronto_venues.shape)
toronto_venues.head()

(1615, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [94]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   74   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             32   
Davisville North                                                       10   
Dufferin, Dovercourt Village                                           17   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         21   
Kensington Market, Chinatown, Grange Park                              59   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             1   
North Toronto West,  Lawrence Park                                     19   
Parkdale, Roncesvalles                                                 15   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     34   
St. James Town                                                         78   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      68   
Christie                                                                16   
Church and Wellesley                                                    74   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              32   
Davisville North                    

In [95]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


In [96]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Theme Restaurant  Toy / Game Store  Trail  \
0             0         0  ...                 0                 0      1   
1             0         0  ...                 0                 0      0   
2             0         0  ...                 0                 0      0   
3             0         0  ...                 0                 0      0   
4             0         0  ...                 0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  
0                      0         0          0              0  
1                      0         0          0              0  
2                      0         0          0              0  
3                      0         0          0              0  
4                      0         0          0              0  

[5 rows x 231 columns]

In [97]:
toronto_onehot.shape

(1615, 231)

In [98]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.014706   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.027027   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.022727   
20                        Moore Park, Summerhill East     0.000000   
21                 North Toronto West,  Lawrence Park     0.052632   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.030303   
24                          Regent Park, Harbourfront     0.022222   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.029412   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010309   
32                                    Studio District     0.025000   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023256   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.028571   

    Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0            0.000000   0.0000              0.0000           0.000   
1            0.000000   0.0000              0.0000           0.000   
2            0.000000   0.0000              0.0000           0.000   
3            0.000000   0.0625              0.0625           0.125   
4            0.000000   0.0000              0.0000           0.000   
5            0.000000   0.0000              0.0000           0.000   
6            0.013514   0.0000              0.0000           0.000   
7            0.000000   0.0000              0.0000           0.000   
8            0.000000   0.0000              0.0000           0.000   
9            0.000000   0.0000              0.0000           0.000   
10           0.000000   0.0000              0.0000           0.000   
11           0.000000   0.0000              0.0000           0.000   
12           0.000000   0.0000              0.0000           0.000   
13           0.000000   0.0000              0.0000           0.000   
14           0.000000   0.0000              0.0000           0.000   
15           0.000000   0.0000              0.0000           0.000   
16           0.000000   0.0000              0.0000

In [99]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.08
1  Cocktail Bar  0.05
2          Café  0.03
3      Beer Bar  0.03
4    Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
                   venue  freq
0                   Café  0.13
1  Performing Arts Venue  0.09
2            Coffee Shop  0.09
3         Breakfast Spot  0.09
4                    Gym  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                venue  freq
0  Light Rail Station  0.12
1         Pizza Place  0.06
2       Auto Workshop  0.06
3          Comic Shop  0.06
4          Restaurant  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.19
1      Airport Lounge  0.12
2    Airport Terminal  0.12
3             Airport  0.06
4  Airport Food Court  0.06


----Central Bay Street----
                 venue  freq
0

In [100]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [101]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0          Cocktail Bar              Beer Bar                   Café   
1           Coffee Shop        Breakfast Spot  Performing Arts Venue   
2           Pizza Place            Skate Park                Brewery   
3        Airport Lounge      Airport Terminal               Boutique   
4    Italian Restaurant        Sandwich Place    Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant                Bakery            Restaurant   
1          Intersection    Italian Restaurant     Convenience Store   
2         Burrito Place                   Spa        Farmers Market   
3       Harbor / Marina         Boat or Ferry   Rental Car Location   
4                  Café       Thai Restaurant                   Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop         Shopping Mall                 Bistro  
1             Nightclub                Bakery              Pet Store  
2  Fast Food Restaurant            Restaurant          Auto Workshop  
3                   Bar                 Plane       Sculpture Garden  
4          Burger Joint       Bubble Tea Shop            Salad Place

In [102]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 4, 0, 4, 4, 0, 4, 4, 0, 0], dtype=int32)

In [103]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_central_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0     Health Food Store                 Trail   
1 -79.352188               4      Greek Restaurant           Coffee Shop   
2 -79.315572               0        Sandwich Place  Fast Food Restaurant   
3 -79.340923               4                  Café           Coffee Shop   
4 -79.388790               0                  Park    Dim Sum Restaurant   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                   Pub         Women's Store            Dance Studio   
1    Italian Restaurant            Restaurant  Furniture / Home Store   
2                  Park      Sushi Restaurant          Ice Cream Shop   
3             Gastropub                Bakery                 Brewery   
4              Bus Line           Swim School    Ethiopian Restaurant   

         6th Most Common Venue        7th Most Common Venue  \
0  Eastern European Restaurant                   Donut Shop   
1                    Bookstore               Ice Cream Shop   
2                   Steakhouse            Fish & Chips Shop   
3          American Restaurant                  Yoga Studio   
4            Electronics Store  Eastern European Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant               Dog Run    Distribution Center  
1                   Pub           Pizza Place                 Lounge  
2               Brewery     Food & Drink Shop     Italian Restaurant  
3     Convenience Store        Sandwich Place            Cheese Shop  
4            Donut Shop      Doner Restaurant                Dog Run

In [104]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters